File: WebApp/indoor-localization/src/modules/report-modules/zone-retention-time-report/managers/ZoneRetentionTimeManager.ts
import { AssetZoneHistoryLog } from "../../../../entities/AssetZoneHistoryLog";

export default class ZoneRetentionTimeManager {
  private dataset: AssetZoneHistoryLog[];

  constructor(dataset: AssetZoneHistoryLog[]) {
    this.dataset = dataset;
  }

  public getAssetIdsInDataset(): number[] {
    return [...new Set(this.dataset.map((log) => log.assetId))];
  }

  public getAssetRetentionTime(assetId: number): number {
    return this.dataset
      .filter((log) => log.assetId === assetId)
      .reduce((sum, log) => sum + (log.retentionTime || 0), 0);
  }

  public getAvgRetentionTime(): number {
    if (this.dataset.length === 0) return 0;

    const totalRetention = this.dataset.reduce(
      (sum, log) => sum + (log.retentionTime || 0),
      0
    );
    return totalRetention / this.dataset.length;
  }

  public getMaxRetentionTime(): number {
    if (this.dataset.length === 0) return 0;

    const assetRetentionTimes = this.getAssetIdsInDataset().map((assetId) =>
      this.getAssetRetentionTime(assetId)
    );

    return Math.max(...assetRetentionTimes);
  }

  public getMinRetentionTime(): number {
    if (this.dataset.length === 0) return 0;

    const assetRetentionTimes = this.getAssetIdsInDataset().map((assetId) =>
      this.getAssetRetentionTime(assetId)
    );

    return Math.min(...assetRetentionTimes);
  }
}

File: Android/modules/zone_retention_time_report/lib/managers/zone_retention_time_manager.dart
class AssetZoneHistoryLog {
  final int assetId;
  final int? retentionTime;

  AssetZoneHistoryLog({required this.assetId, this.retentionTime});
}

class ZoneRetentionTimeManager {
  final List<AssetZoneHistoryLog> dataset;

  ZoneRetentionTimeManager(this.dataset);

  List<int> getAssetIdsInDataset() {
    return dataset.map((log) => log.assetId).toSet().toList();
  }

  int getAssetRetentionTime(int assetId) {
    return dataset
        .where((log) => log.assetId == assetId)
        .fold(0, (sum, log) => sum + (log.retentionTime ?? 0));
  }

  double getAvgRetentionTime() {
    if (dataset.isEmpty) return 0;

    final totalRetention = dataset.fold(
        0, (sum, log) => sum + (log.retentionTime ?? 0));
    return totalRetention / dataset.length;
  }

  int getMaxRetentionTime() {
    if (dataset.isEmpty) return 0;

    final assetRetentionTimes = getAssetIdsInDataset()
        .map((assetId) => getAssetRetentionTime(assetId))
        .toList();

    return assetRetentionTimes.reduce((a, b) => a > b ? a : b);
  }

  int getMinRetentionTime() {
    if (dataset.isEmpty) return 0;

    final assetRetentionTimes = getAssetIdsInDataset()
        .map((assetId) => getAssetRetentionTime(assetId))
        .toList();

    return assetRetentionTimes.reduce((a, b) => a < b ? a : b);
  }
}